patches/tModLoader/Terraria/ModLoader/LocalizationLoader.cs

```diff
@@ -286,4 +295,317 @@ private static LocalizedText SetLocalizedText(Dictionary<string, LocalizedText>
+
+			allLocalizationFilesAllLanguages.Add(translationFile.Name);
+			allLanguages.Add(culture);
+
+			// TODO: Support arbitrary default language?
+			// Default language hjson files loaded into memory to gather comments and modder intended ordering.
+			if (culture == GameCulture.FromCultureName(GameCulture.CultureName.English)) {
+				JsonValue jsonValueEng = HjsonValue.Parse(translationFileContents, new HjsonOptions() { KeepWsc = true });
+				JsonObject jsonObjectEng = jsonValueEng.Qo();
+				// Default language files are flattened to a different data structure here to avoid confusing WscJsonObject manipulation with Prefix.AnotherPrefix-type keys and comment preservation.
+				List<LocalizationEntry> existingEntries = FlattenJsonObject(jsonObjectEng as WscJsonObject, prefix);
+				localizationFileEntries.Add(new(translationFile.Name, prefix, existingEntries));
+			}
+		}
+
+		// Abort if no default localization files found
+		if (!localizationFileEntries.Any())
+			return;
+
+		var existingKeys = new List<LocalizationEntry>();
+		// Collect known keys. These are potentially missing from the localization files
+		foreach (var translation in translations) {
+			if (translation.Key.StartsWith($"Mods.{mod.Name}.")) {
+				existingKeys.Add(new(translation.Key, translation.Value.GetDefault(), null));
+			}
+		}
+
+		// Merge collected keys into flattened in-memory model
+		foreach (var existingKey in existingKeys) {
+			LocalizationFileEntry suitableHJSONFile = FindHJSONFileForKey(localizationFileEntries, existingKey.key);
+
+			if (!KeyExistsInHJSON(suitableHJSONFile, existingKey.key)) {
+				AddEntryToHJSON(suitableHJSONFile, existingKey.key, existingKey.value, null);
+			}
+			// What do we do if the hjson and loaded translations are different, will that be possible? Check Override here once implemented
+		}
+
+		if(specificCulture != null) {
+			allLanguages.Clear();
+			allLanguages.Add(specificCulture);
+		}
+
+		// Update all languages that have been found in the mod
+		foreach (var culture in allLanguages) {
+			// Save all localization files
+			foreach (var localizationFileEntry in localizationFileEntries) {
+				// TODO: Don't save to disk if unchange optimization
+
+				var rootObject = new CommentedWscJsonObject();
+				// Convert back to JsonObject and write to disk
+				for (int i = 0; i < localizationFileEntry.LocalizationEntries.Count; i++) {
+					var entry = localizationFileEntry.LocalizationEntries[i];
+
+					CommentedWscJsonObject parent = rootObject;
+
+					string key = entry.key;
+					if (!string.IsNullOrWhiteSpace(localizationFileEntry.prefix)) {
+						key = key.Substring(localizationFileEntry.prefix.Length + 1);
+					}
+
+					// Find/Populate the parents of this translation entry 
+					string[] splitKey = key.Split(".");
+					for (int j = 0; j < splitKey.Length - 1; j++) {
+						string k = splitKey[j];
+						if (parent.ContainsKey(k))
+							parent = (CommentedWscJsonObject)parent[k];
```

Comment:
Calling back to slack discussion. What if one of the 'new' translations (from a new piece of content I just added) has a parent which already has a value? It will be a `JsonValue` rather than a `CommentedWscJsonObject`

We would need to support `"$parentVal"` or a preferred alternative

Reply:
I can't see a way around a special key like $parentVal, other than not allowing keys like that. I'm going to hold off on implementing this one while we focus on getting everything else working.